### PR TITLE
feat: no_sig_flattening option

### PR DESCRIPTION
**Pull request by guybedford**
Was exploring https://github.com/bytecodealliance/wit-bindgen/issues/524 today, and was considering if we should disable all flattening by default as well.

This PR implements a `no_sig_flattening` option for the C generation as suggested.

I did explore entirely removing the flattening based on the argument that one might want a more "low level" interface, but at the end of the day even the C interface is a high level interface! And if anything, the flattening is actually slightly more optimal from a bindings perspective (while trivial, it does avoid unnecessary allocations of stack objects in lifted interfaces).

So I think having this option may well just be the best approach here.

This is just a draft initial exploration of the work, not ready to merge.

### CODE REVIEW

File: crates/c/src/lib.rs [hunk 1]
```diff
@@ -35,6 +35,9 @@ pub struct Opts {
     /// Set component string encoding
     #[cfg_attr(feature = "clap", arg(long, default_value_t = StringEncoding::default()))]
     pub string_encoding: StringEncoding,
+    // Skip optional null pointer and boolean result argument signature flattening
+    #[cfg_attr(feature = "clap", arg(long, default_value_t = false))]
+    pub no_sig_flattening: bool,
```
Comment (Mossaka) [inline crates/c/src/lib.rs:40]: Would it be more precise and descriptive to name this flag as `no_option_ty_flattening`?

File: crates/c/src/lib.rs [hunk 2]
```diff
@@ -466,22 +475,26 @@ impl Return {
...
 
             // Unpack a result as a boolean return type, with two
             // return pointers for ok and err values
             TypeDefKind::Result(r) => {
-                if let Some(ok) = r.ok {
-                    self.retptrs.push(ok);
-                }
-                if let Some(err) = r.err {
-                    self.retptrs.push(err);
+                if sig_flattening {
```
Comment (Mossaka) [inline crates/c/src/lib.rs:488]: Nice!

**Comment by Mossaka** [comment]
For `no_sig_flattening = true`, can we remove all the option flattening for both arguments and results?

I am still seeing the optional arguments being flattened even if the flag is set to true. 

```wit
interface option-result {
	type o1 = option<u32>
    o1-arg: func(x: o1)
    o1-result: func() -> o1

    record empty {}
    type o2 = option<empty>
    o2-arg: func(x: o2)
    o2-result: func() -> o2
}

default world my-world {
	import imports: self.option-result
	export exports: self.option-result
}
  
```
```C
// Imported Functions from `imports`
void imports_o1_arg(uint32_t *maybe_x);
void imports_o1_result(imports_o1_t *ret);
void imports_o2_arg(imports_empty_t *maybe_x);
void imports_o2_result(imports_o2_t *ret);

// Exported Functions from `exports`
void exports_o1_arg(uint32_t *maybe_x);
void exports_o1_result(exports_o1_t *ret);
void exports_o2_arg(exports_empty_t *maybe_x);
void exports_o2_result(exports_o2_t *ret);
```

## Expected

```C
```

**Comment by Mossaka** [comment]
I pushed a commit to fix the above issue.

**Comment by Mossaka** [comment]
All the runtime and codegen tests for go bindgen are passing for me.

**Comment by alexcrichton** [comment]
@Mossaka would you be up for rebasing this and handling my comment above?

**Comment by Mossaka** [comment]
> @Mossaka would you be up for rebasing this and handling my comment above?

Yes, WIP I am trying to enable all the codegen tests for go bindgen. 

What comment were you referring to? I didn't see you have any comment above.

**Comment by Mossaka** [comment]
I've enable all the codegen go tests except issue 544 and issue 551. I intend to follow up with another PR to resolve these two tests. For now, I dis-abled them in the codegen tests, and thus this PR is ready to review!

**Review by guybedford — COMMENTED** [review]
LGTM!

**Review by pchickey — APPROVED** [review]

**Pull request merged**